### PR TITLE
Fixed: GetArchiveLink now accepts 301 redirection

### PR DESCRIPTION
**Pull request by vaibhavsingh97**
fixes: #1248

### CODE REVIEW

**Comment by vaibhavsingh97** [comment]
@gmlewis please review 😄

File: github/repos_contents.go [hunk 1]
```diff
@@ -261,7 +261,7 @@ func (s *RepositoriesService) GetArchiveLink(ctx context.Context, owner, repo st
 		return nil, nil, err
 	}
 	resp.Body.Close()
-	if resp.StatusCode != http.StatusFound {
+	if !(resp.StatusCode == http.StatusFound || resp.StatusCode == http.StatusMovedPermanently) {
```
Comment (gmlewis) [inline github/repos_contents.go]: How about: 

`if resp.StatusCode != http.StatusFound && resp.StatusCode != http.StatusMovedPermanently {`

here and in the test file?

**Comment by Soulou** [comment]
I've actually another point : 301 redirection happens when a repository has been renamed, so isn't there one more rountrip to do to get the archive ?

`https://api.github.com/user/repo/archive?ref=master` --- 301 ---> `https://api.github.com/user/new-name/archive?ref=master` --- 302 ---> `https://codeload.github.com/user/new-name/archive?ref=master&token=<token>?

**Comment by gmlewis** [comment]
Yes, your client will need to recognize and follow the link if it wants it.
As I mentioned before, we are not going to have this package automatically follow redirects... it will be up to the user of this package to detect the redirect and perform whatever action it wants to perform.

**Comment by gmlewis** [comment]
@vaibhavsingh97 - can you please verify that the redirect URL will be available to the user of this package so it can take action, and add that URL value to the unit test as a demonstration of usage?

**Comment by Soulou** [comment]
@gmlewis In the case the 301 redirect would target the API requiring authentication, it's not possible and the package has to follow the redirection until we got the `Location` with the authenticated token

**Comment by gmlewis** [comment]
The reason we do it this way, @Soulou is so that the user of this package has full control over what they want to do.

I can imagine that some program may wish to move a bunch of repos, and then use this API to poll GitHub periodically until the redirect is properly returned, and then report back that the moves have successfully completed (without actually needing to follow the redirect).

**Comment by gmlewis** [comment]
@Soulou - I'm not sure I understand what you are saying is not possible.
If this API returns the URL along with the 301 in the response, then your program can then get the new URL and then make another request with the new URL to get the repo info.

**Comment by Soulou** [comment]
@gmlewis what I mean is the following:

Without any renaming:

```
└> curl --user ":$TOKEN" https://api.github.com/repos/user/repo/tarball/master -v -q 2>&1 | grep Location
< Location: https://codeload.github.com/repo/user/legacy.tar.gz/master?token=AACQR4JB4AR5CSOK5KLWXG
```

The API returns an URL which is directly usable by the client of the package `google/go-github`, it's authentified by a token valid for 5 minutes (according to GitHub documentation)

Once the app has been renamed, 301 is returned with the following

```
└> curl --user ":$TOKEN" https://api.github.com/repos/user/renamed/tarball/master -v -q 2>&1 | grep Location
< Location: https://api.github.com/repositories/<ID>/tarball/master
```

So the user of this package would have to check if:

* It is an URL which can be used to download the tarball itself
* It is still an URL targeting the API which does not contain all the information to make another call again: (indeed we don't have the new name here, and the name is required in this package if I'm not mistaken):

```
	client.RepositoriesService.GetArchiveLink(
		ctx, owner, repo, github.Tarball,
		&github.RepositoryContentGetOptions{Ref: ref},
	)
```

I think it would be really a lot of responsability for the user of this package, while if an additional roundtrip would be done here, it would be much easier and logical. `GetArchiveLink` would always return an archive link as defined in https://developer.github.com/v3/repos/contents/#get-archive-link and not a link to analyze.

**Comment by gmlewis** [comment]
Thank you for the explanation, @Soulou!

That makes a lot of sense, and this PR will not be sufficient to fix #1248.

Since only two lines were slightly modified here, I'm going to close this PR and let's start over on this one.  I'll also copy your detailed description to the desired solution into the original issue.

**Pull request closed without merging**